tests/unit/test_micro_agents.py

```diff
@@ -49,8 +47,8 @@ def test_coder_agent_with_summary():
     state = State(history=history, inputs={'summary': summary})
     coder_agent.step(state)
 
-    mock_llm.do_completion.assert_called_once()
-    _, kwargs = mock_llm.do_completion.call_args
+    mock_llm.completion.assert_called_once()
+    _, kwargs = mock_llm.completion.call_args
```

Comment:
"completion" itself has no parameters anymore. Does this test with "call_args" make still sense (it's a mock, I know) or would it now look at the wrapped params?

Reply:
I'm looking at the llm.py class, and `completion` is `self._completion`, which is assigned the `wrapper(*args, **kwargs)` method. Just like it was assigned to `do_completion`, and just like it was assigned to `completion` prior to that.

If it accepts no parameters in this PR, then that's absolutely not intended. But why do you think it doesn't?

Reply:
Right, I was looking at this:
![grafik](https://github.com/OpenDevin/OpenDevin/assets/10787084/2bfc01c2-7f0a-42f5-8683-2c212aab3578)

Before, all calls were to `do_completion`, which had args/kwargs. This is now removed and is now just `completion` but with parameters. It's more an educational question to learn the details (coming from C#): no the passed params for it refer directly to the wrapped parameters, but the method itself doesn't have a signature for parameters anymore.

Reply:
Hehe, I find its full story a bit difficult to follow too. 😅 But the short answer on your screenshot is: because we sent in the @property a reference to the function. That's not a function call. `return self._completion` is not `return._completion()` so it's not a call, it didn't 'skip' any args, it's a pointer to the function stored in `_completion`, with all its stuff.

Try it just for fun: assign something before return, like `decorated_function = self._completion`, then print() its type (it will be function type), its __dict__, etc.

Here's a quick screenshot, I stopped it in the debugger at the first LLM call: this is the moment it enters litellm, after it passed through our `completion` decorator and everything else. On the right, the kwargs are what we'd expect: messages, temperature, and stop.
![Screenshot 2024-06-19 at 01 22 11](https://github.com/OpenDevin/OpenDevin/assets/6080905/77f8f4e7-765f-49ff-bbb0-2e5f2083d268)

Reply:
About the fact that `completion(self)` doesn't, itself, have args: it's a property, a getter, it just gives access to the attribute. In this case, the attribute is a function object. When the code does `llm.completion(arg1, arg2, kwarg1=something, or, whatever)`, the python interpreter retrieves the underlying object (the attribute value) and it invokes it with those parameters.

Reply:
That clarifies it really well, thank you!